src/snowflake/mod.rs

```diff
@@ -92,6 +105,93 @@ impl object_store::CredentialProvider for S3StageCredentialProvider {
 }
 
 
+#[derive(Debug)]
+pub(crate) struct SnowflakeAzureExtension {
+    stage: String,
+    client: Arc<SnowflakeClient>,
+}
+
+#[async_trait::async_trait]
+impl Extension for SnowflakeAzureExtension {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+    async fn current_stage_info(&self) -> crate::Result<String> {
+        let stage_info = &self
+            .client
+            .current_upload_info(&self.stage)
+            .await?
+            .stage_info;
+        let stage_info: NormalizedStageInfo = stage_info.try_into()?;
+        let string = serde_json::to_string(&stage_info)
+            .context("failed to encode stage_info as json").to_err()?;
+        Ok(string)
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct AzureStageCredentialProvider {
+    stage: String,
+    client: Arc<SnowflakeClient>,
+    cached: Mutex<Option<Arc<object_store::azure::AzureCredential>>>
+}
+
+impl AzureStageCredentialProvider {
+    pub(crate) fn new(stage: impl AsRef<str>, client: Arc<SnowflakeClient>) -> AzureStageCredentialProvider {
+        AzureStageCredentialProvider { stage: stage.as_ref().to_string(), client, cached: Mutex::new(None) }
+    }
+}
+
+#[async_trait::async_trait]
+impl object_store::CredentialProvider for AzureStageCredentialProvider {
+    type Credential = object_store::azure::AzureCredential;
+    async fn get_credential(&self) -> object_store::Result<Arc<Self::Credential>> {
+        let info = self.client.current_upload_info(&self.stage).await
+            .map_err(|e| {
+                object_store::Error::Generic {
+                    store: "MicrosoftAzure",
+                    source: e.into()
+                }
+            })?;
+
+        if info.stage_info.location_type != "AZURE" {
+            return Err(object_store::Error::Generic {
+                store: "MicrosoftAzure",
+                source: Error::invalid_response("Location type must be AZURE for this provider").into()
+            })
+        }
+
+        let new_creds = info.stage_info.creds.as_azure()
+            .map_err(|e| object_store::Error::Generic {
+                store: "MicrosoftAzure",
+                source: e.into()
+            })?;
+        
+
+        let token_bytes = new_creds.azure_sas_token.trim_start_matches('?').as_bytes();
+        let new_pairs = url::form_urlencoded::parse(token_bytes)
```

Comment:
I was not aware that `url` re-exports `form_urlencoded`! That looks much cleaner than the hack on my version.